File: src/org/kde/kdeconnect/Plugins/MprisPlugin/MprisMediaNotificationReceiver.java
/*
 * SPDX-FileCopyrightText: 2017 Matthijs Tijink <matthijstijink@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/

package org.kde.kdeconnect.Plugins.MprisPlugin;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.support.v4.media.session.MediaSessionCompat;

import org.kde.kdeconnect.BackgroundService;
import org.kde.kdeconnect.Device;

/**
 * Called when the mpris media notification's buttons are pressed
 */
public class MprisMediaNotificationReceiver extends BroadcastReceiver {
    public static final String ACTION_PLAY = "ACTION_PLAY";
    public static final String ACTION_PAUSE = "ACTION_PAUSE";
    public static final String ACTION_PREVIOUS = "ACTION_PREVIOUS";
    public static final String ACTION_NEXT = "ACTION_NEXT";
    public static final String ACTION_CLOSE_NOTIFICATION = "ACTION_CLOSE_NOTIFICATION";

    public static final String EXTRA_DEVICE_ID = "deviceId";
    public static final String EXTRA_MPRIS_PLAYER = "player";

    @Override
    public void onReceive(Context context, Intent intent) {
        //First case: buttons send by other applications via the media session APIs
        if (Intent.ACTION_MEDIA_BUTTON.equals(intent.getAction())) {
            //Route these buttons to the media session, which will handle them
            MediaSessionCompat mediaSession = MprisMediaSession.getMediaSession();
            if (mediaSession == null) return;
            mediaSession.getController().dispatchMediaButtonEvent(intent.getParcelableExtra(Intent.EXTRA_KEY_EVENT));
        } else {
            //Second case: buttons on the notification, which we created ourselves

            //Get the correct device, the mpris plugin and the mpris player
            BackgroundService service = BackgroundService.getInstance();
            if (service == null) return;
            Device device = service.getDevice(intent.getStringExtra(EXTRA_DEVICE_ID));
            if (device == null) return;
            MprisPlugin mpris = device.getPlugin(MprisPlugin.class);
            if (mpris == null) return;
            MprisPlugin.MprisPlayer player = mpris.getPlayerStatus(intent.getStringExtra(EXTRA_MPRIS_PLAYER));
            if (player == null) return;

            //Forward the action to the player
            switch (intent.getAction()) {
                case ACTION_PLAY:
                    player.play();
                    break;
                case ACTION_PAUSE:
                    player.pause();
                    break;
                case ACTION_PREVIOUS:
                    player.previous();
                    break;
                case ACTION_NEXT:
                    player.next();
                    break;
                case ACTION_CLOSE_NOTIFICATION:
                    //The user dismissed the notification: actually handle its removal correctly
                    MprisMediaSession.getInstance().closeMediaNotification();
            }
        }
    }
}

File: src/main/java/org/kde/kdeconnect/plugins/mpris/MprisMediaNotificationReceiver.kt
/*
 * SPDX-FileCopyrightText: 2017 Matthijs Tijink <matthijstijink@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/
package org.kde.kdeconnect.plugins.mpris

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log
import org.kde.kdeconnect.KdeConnect
import org.kde.kdeconnect.extensions.getParcelableCompat

/**
 * Called when the mpris media notification's buttons are pressed
 */
class MprisMediaNotificationReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        // First case: buttons send by other applications via the media session APIs. They don't target a specific device.
        if (Intent.ACTION_MEDIA_BUTTON == intent.action) {
            // Route these buttons to the media session, which will handle them
            val mediaSession = MprisMediaSession.getMediaSession() ?: return
            mediaSession.controller.dispatchMediaButtonEvent(intent.getParcelableCompat(Intent.EXTRA_KEY_EVENT))
        } else {
            // Second case: buttons on the notification, which we created ourselves
            // Get the correct device, the mpris plugin and the mpris player

            val deviceId = intent.getStringExtra(EXTRA_DEVICE_ID)
            val plugin = KdeConnect.getInstance().getDevicePlugin(deviceId, MprisPlugin::class.java) ?: return
            val player = plugin.getPlayerStatus(intent.getStringExtra(EXTRA_MPRIS_PLAYER))
                ?: return

            when (intent.action) {
                ACTION_PLAY -> player.sendPlay()
                ACTION_PAUSE -> player.sendPause()
                ACTION_PREVIOUS -> player.sendPrevious()
                ACTION_NEXT -> player.sendNext()
                ACTION_CLOSE_NOTIFICATION ->                     //The user dismissed the notification: actually handle its removal correctly
                    MprisMediaSession.instance.closeMediaNotification()
                else -> {
                    Log.w(TAG, "Unknown action: ${intent.action}, ignore.")
                }
            }
        }
    }

    companion object {
        const val TAG: String = "M.M.N.Receiver"

        const val ACTION_PLAY: String = "ACTION_PLAY"
        const val ACTION_PAUSE: String = "ACTION_PAUSE"
        const val ACTION_PREVIOUS: String = "ACTION_PREVIOUS"
        const val ACTION_NEXT: String = "ACTION_NEXT"
        const val ACTION_CLOSE_NOTIFICATION: String = "ACTION_CLOSE_NOTIFICATION"

        const val EXTRA_DEVICE_ID: String = "deviceId"
        const val EXTRA_MPRIS_PLAYER: String = "player"
    }
}
